### PR TITLE
#669-Bug: Analog time doesn't correspond with digital time

### PR DESCRIPTION
**Pull request by HeinKhantZaw**
## Description

Issue #669 is fixed. I fixed the wrong time error for both analog clock and digital clock in world clock utility. Now, the clock works correctly whatever the device's local time is (for both 12 hour format and 24 hour format).

### **Before Fix**
before

### **After Fix**
after

## Type of change
Just put an x in the [] which are valid.
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
- [x] `./gradlew assembleDebug assembleRelease`
- [x] `./gradlew checkstyle`

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        super(context, attrs);
+        handleAttrs(context, attrs);
+    }
+
+    public CustomAnalogClock(Context context) {
+        super(context);
+        init(context);
+    }
+
+    public CustomAnalogClock(Context context, boolean defaultWatchFace) {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: add javadoc style comments to all the newly added functions

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java [hunk 2]
```diff
@@ -2,44 +2,36 @@
...
-
-    //private DigitalClock mDigitalClock;
-
-    private static TextClock mDigitalClock;
+    private CustomAnalogClock mAnalogClock;
+    private static final String DEFAULT_TIME_ZONE_KEY = "defaultTimeZone";
+    private TextClock mTextClock;
+    private long mMiliSeconds;
+    private ArrayAdapter<String> mIdAdapter;
+    private Spinner mSpinnerAvailableID;
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: what does this spinner indicate? It's not clear from name

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java [hunk 3]
```diff
@@ -118,19 +97,40 @@ public boolean onOptionsItemSelected(MenuItem item) {
         return super.onOptionsItemSelected(item);
     }
 
-
-    /**
-     * Pulls 12/24 mode from system settings
-     */
-    private boolean get24HourMode() {
-        return DateFormat.is24HourFormat(mContext);
+    private void getGMTTime() {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: add javadoc style comments to newly added functions

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java [hunk 4]
```diff
@@ -118,19 +97,40 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+        mMiliSeconds = mMiliSeconds - offset;
     }
 
-    private void setFormat() {
-        if (get24HourMode()) {
-            mDigitalClock.setFormat24Hour(mHours24);
-        } else {
-            mDigitalClock.setTimeZone(mHours12);
-        }
+    private String getSavedValue(String key) {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: no need of a new function just for this purpose

File: Android/app/src/main/res/layout/activity_utilities_world_clock.xml [hunk 5]
```diff
@@ -1,86 +1,75 @@
...
+                android:layout_marginTop="8dp"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toBottomOf="@+id/tvTimeZone">
 
-                <RelativeLayout
-                    android:id="@+id/zone_field"
-                    android:layout_width="match_parent"
+                <Spinner
+                    android:id="@+id/availableID"
```
Comment (Swati4star) [inline Android/app/src/main/res/layout/activity_utilities_world_clock.xml]: what is this spinner for?

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java [hunk 6]
```diff
@@ -118,19 +97,40 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+        // this code get the latest time zone that user have chosen
+        return PreferenceManager.getDefaultSharedPreferences(getBaseContext())
+                .getString(key, mDefaultZone);
     }
+    private void saveTimeZonePrefs(String key, String value) {
+        // this code is for saving the user selected time zone
+        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(getBaseContext());
+        settings.edit().putString(key, value).apply();
+    }
+    private void setSelectedText(TimeZone timezone) {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: maintain the formatting

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java (hunk 1 above)
Comment (HeinKhantZaw) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: Added

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 2 above)
Comment (HeinKhantZaw) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: This spinner shows the available time zones for the users.

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 4 above)
Comment (HeinKhantZaw) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: Shall I remove it?

File: Android/app/src/main/res/layout/activity_utilities_world_clock.xml (hunk 5 above)
Comment (HeinKhantZaw) [reply Android/app/src/main/res/layout/activity_utilities_world_clock.xml]: This spinner is used to show the available time zones.

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java (hunk 1 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: not added for all yet

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        typedArray.recycle();
+    }
+
+    public void init(Context context) {
+        init(context, R.drawable.clock_face, R.drawable.hours_hand,
+                R.drawable.minutes_hand, 0, false, false);
+    }
+
+    /**
+     * Will set the scale of the view, for example 0.5f will draw the clock with half of its mRadius
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: remove "will set"
change it to "Sets"

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    public void setFace(int drawableRes) {
+        final Resources r = getResources();
+        setFace(r.getDrawable(drawableRes));
+    }
+
+    /**
+     * Initialize the analog clock
+     */
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: not properly formatted

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        if (alpha > 0)
+            hHand.setAlpha(alpha);
+
+        Drawable mHand = context.getResources().getDrawable(minuteHand);
+
+        mCalendar = Calendar.getInstance();
+
+        mHandsOverlay = new HandsOverlay(hHand, mHand).withScale(mSizeScale);
+    }
+    /**
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: add new line after functions

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/HandsOverlay.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+    private final Drawable mHour;
+    private final Drawable mMinute;
+    private final boolean mUseLargeFace;
+    private float mHourRot;
+    private float mMinRot;
+    private boolean mShowSeconds;
+    private float mScale;
+
+    public HandsOverlay(Context context, boolean useLargeFace) {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/HandsOverlay.java]: add comments here also, please.

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 2 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: not clear from var name. make it verbose

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 3 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: not done yet

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 4 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: yes

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java (hunk 6 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/WorldClockActivity.java]: not done yet

**Comment by HeinKhantZaw** [comment]
I am not familiar with OSS contributions. That's why I wanna say sorry for my mistakes. I am willing to change my PR again if you want me to change. Thanks in advanced for reviewing my PR.

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java (hunk 1 above)
Comment (Swati4star) [reply Android/app/src/main/java/io/github/project_travel_mate/utilities/CustomAnalogClock.java]: ??

File: Android/app/src/main/java/io/github/project_travel_mate/utilities/HandsOverlay.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        final Resources r = context.getResources();
+
+        mUseLargeFace = useLargeFace;
+
+        mHour = null;
+        mMinute = null;
+    }
+
+    public HandsOverlay(Drawable hourHand, Drawable minuteHand) {
```
Comment (Swati4star) [inline Android/app/src/main/java/io/github/project_travel_mate/utilities/HandsOverlay.java]: I mean for all newly added functions, Not just the one I add comment to

**Comment by HeinKhantZaw** [comment]
I have added comments to every function.

**Comment by Swati4star** [comment]
Great job working through so many iterations! @HeinKhantZaw :)

**Comment by HeinKhantZaw** [comment]
Sorry for making you busier too. Although this is my first time for contributing OSS,thank you so much for guiding me @Swati4star. I am willing to contribute to your project more. ;)

**Comment by Swati4star** [comment]
Awesome @HeinKhantZaw :)

**Pull request merged**